pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -244,29 +245,45 @@ func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus
 	return model.UnHealthy
 }
 
-func (esc *endpointSliceController) updateEndpointCacheForSlice(hostName host.Name, slice *v1.EndpointSlice) {
+func (esc *endpointSliceController) updateEndpointCacheForSlice(hostName host.Name, epSlice *v1.EndpointSlice) {
 	var endpoints []*model.IstioEndpoint
-	if slice.AddressType == v1.AddressTypeFQDN {
+	if epSlice.AddressType == v1.AddressTypeFQDN {
 		// TODO(https://github.com/istio/istio/issues/34995) support FQDN endpointslice
 		return
 	}
 	svc := esc.c.GetService(hostName)
-	svcNamespacedName := getServiceNamespacedName(slice)
+	svcNamespacedName := getServiceNamespacedName(epSlice)
 	svcCore := esc.c.services.Get(svcNamespacedName.Name, svcNamespacedName.Namespace)
+	if svcCore != nil && len(svcCore.Spec.ClusterIPs) > 1 {
+		// It means this is dual stack service, for which k8s will build two endpointslice per pod.
```

Comment:
I think its possible to have a dualstack service which only has 1 IPv6 address. I doubt kubernetes endpoint controller will generate it, but users can create endpoints by hand and we need to support those.

Reply:
Ah good catch, users could make a service without a selector and then manually create only an IPv6 EndpointSlices

Reply:
Help me understand, how could dual stack service have one clusterIP?

Reply:
I do not mean 1 ClusterIp, but only 1 endpointslice with only ipv6

Reply:
Per k8s, i think if the service is ipv6, it will by default generate iptables with only ipv6. But for ipv4, it generated iptables for ipv4. 

For the case, user created endpointslice themself,  i think they must obet the service ip family. I feel if a service with ipv6 family, but user create endpointslice with ipv4 address, they could not access.

So what is missing for this case, i am not quite sure

Reply:
I don't think there is anything suggesting a dual stack family has to have a pair of v4-v6 endpoints for everything it's representing. There is no correlation between the two - they are two entirely orthogonal sets of endpoints. 

Istio is artificially joining the endpoints with pods based on a heuristic (for general use, not just dual stack), and now for dual stack we are *joining* these based on the pod. 

But this PR does not do a full join, it only uses the v4 on. So if there is no v4 at all, which is valid, we miss it. 

These slices may not even be bound to pods at all, in which case this is even more broke since there is no "join" happening. In that case we just lose all of the Ipv6 endpoints entirely since we couldn't recover them. 

Obviously we have a huge testing gap for these edge cases.. happy to help expand those

Reply:
TBH the pod case is probably not going to happen in real world, but the non-pod case is just dropping all the Ipv6 endpoints entirely here which is not even an edge case

Reply:
Not sure that made sense, I can provide a concrete test case when I'm at a computer if it will help

Reply:
I expect a concrete example to be more clear

Reply:
BTW, the below check does not mean, we only generate we ignore ipv6, becasue we acquire the ips for pod at last. So all ips in pod status will be there. 

```
if epSlice.AddressType == v1.AddressTypeIPv6 {
			return
		}
```

Reply:
I am mostly worried about endpoints that are not for Pods at all.

Reply:
IC, indeed missed processing